File: control-plane/src/server/services/triage.ts
import type { FleetEvent } from "@shared/events";
import { wakesFleetManager } from "@shared/events";

export class Triage {
  private recentDups: { workflowId: string; at: number }[] = [];

  shouldWake(e: FleetEvent): boolean { return wakesFleetManager(e); }

  observe(e: FleetEvent, now: number = Date.now()): void {
    if (e.type === "workflow.exception.detected" && e.category === "duplicate-invoice") {
      this.recentDups.push({ workflowId: e.workflowId, at: now });
      this.recentDups = this.recentDups.filter(r => now - r.at <= 60_000);
    }
  }

  detectAnomaly(now: number = Date.now()): { pattern: string; workflowIds: string[] } | null {
    const dups = this.recentDups.filter(r => now - r.at <= 60_000);
    if (dups.length >= 3) {
      return { pattern: "duplicate-burst", workflowIds: dups.map(d => d.workflowId) };
    }
    return null;
  }
}

File: api/server/services/triage.py
from __future__ import annotations
import time
from api.shared.events import FleetEvent, wakes_fleet_manager
from api.shared import domains as _registry


class Triage:
    def __init__(self) -> None:
        self._recent_dups: list[tuple[str, float]] = []

    def should_wake(self, e: FleetEvent) -> bool:
        # Phase 4 of feature-fleet-domain-substrate-1: in addition to the
        # platform-wide WAKE_TYPES (workflow.exception.detected, hitl,
        # SLA breach, anomaly, tick, claim.routed.red), wake the FM on
        # any per-domain wake hint declared in the registry.
        if wakes_fleet_manager(e):
            return True
        return e.type in _registry.all_wake_hints()

    def observe(self, e: FleetEvent, now: float | None = None) -> None:
        now = now if now is not None else time.time()
        if e.type == "workflow.exception.detected" and getattr(e, "category", None) == "duplicate-invoice":
            self._recent_dups.append((e.workflow_id or "", now))
            self._recent_dups = [(w, t) for w, t in self._recent_dups if now - t <= 60]

    def detect_anomaly(self, now: float | None = None) -> dict | None:
        now = now if now is not None else time.time()
        dups = [(w, t) for w, t in self._recent_dups if now - t <= 60]
        if len(dups) >= 3:
            return {"pattern": "duplicate-burst", "workflow_ids": [w for w, _ in dups]}
        return None
